keylime/elchecking/tests.py

```diff
@@ -0,0 +1,594 @@
+import abc
+import re
+import typing
+
+# This module defines the abstraction of a Test (of JSON data)
+# and several specific test classes.
+# A Test can be used multiple times, even concurrently.
+
+# Data is the type of Python data that corresponds to JSON values.
+Data = typing.Union[int, float, str, bool, typing.Tuple['Data', ...],
+                    typing.Mapping[str, 'Data'], None]
+
+# Globals is a dict of variables for communication among tests.
+# There is a distinct dict for each top-level use of a test.
+Globals = typing.Mapping[str, Data]
+
+# PCR_Contents maps digest name to map from PCR index to PCR value.
+# Here digest name is something like 'sha256'.
+# Each PCR index is a decimal string, so that this can be JSON data
+PCR_Contents = typing.Mapping[str, typing.Mapping[str, int]]
+
+
+class Test(metaclass=abc.ABCMeta):
+    """Test is something that can examine a value and either approve it or give a reason for rejection"""
+
+    @abc.abstractmethod
+    def why_not(self, globs: Globals, subject: Data) -> str:
+        """Test the given value, return empty string for pass, explanation for fail.
+
+        The explanation is (except in deliberate exceptions) English that
+        makes a sentence when placed after a noun phrase identifying the subject.
+        The test can read and write in the given globs dict.
+        """
+        raise NotImplementedError
+
+
+# type_test constructs a test of data type that is expected to pass
+def type_test(t) -> typing.Callable[[typing.Any], bool]:
+    """Returns a lambda that tests against the given type.
+
+    The lambda returns True on pass, raises Exception on fail."""
+    def test(v: typing.Any) -> bool:
+        if isinstance(v, t):
+            return True
+        raise Exception(f'{v!r} is a {type(v)} rather than a {t}')
+    return test
+
+
+def list_test(elt_test: typing.Callable[[typing.Any], bool]) -> typing.Callable[[typing.Any], bool]:
+    """Return a lambda that tests for list with certain type of element"""
+    def test(dat: typing.Any) -> bool:
+        type_test(list)(dat)
+        for elt in dat:
+            elt_test(elt)
+        return True
+    return test
+
+
+def dict_test(dom_test: typing.Callable[[typing.Any], bool], rng_test: typing.Callable[[typing.Any], bool]) -> typing.Callable[[typing.Any], bool]:
+    """Return a lambda that tests for dict with certain type key and value"""
+    def test(dat: typing.Any) -> bool:
+        type_test(dict)(dat)
+        for dom, rng in dat.items():
```

Comment:
I'm curious about the naming here. What are `dom` and `rng` references to and why use them instead of something more familiar like `key`/`value` ?

Reply:
@MikeSpreitzer can you comment on that?

Reply:
`dom` and `rng` are short for `domain` and `range`, terms I thought were common for the two sides of a function (be it finite or otherwise).  If you think it matters, these can easily be changed to `key` and `value`.

Reply:
You're right those are common names for the 2 sides of a function, it just wasn't obv to me that's what we were testing. The method is called `dict_test` and the comment mentions testing the dict for a certain key and value. That's pretty generic and wouldn't necessarily be fore testing the the domain and range of another function. Also because it wasn't immediately obv to me this was what was happening, `rng` looked like `RNG` or random-number generator and I was confused even more :)

Unless there's something specific about testing a domain and range here (which I don't see, but might be more obv in how it's used) I would suggest something more generic like `key`/`value` or `k`/`v`.

Reply:
@mpeters I proposed we merge as is, and we will fix it on the "PR5", the one which will also include measured boot-specific tests, as we previously discussed.